---
title: "Bulk_workflow"
author: "Stefania Pirrotta"
output: html_document
---

# Prepare Data
In this workflow, we present all the analyses, along with the code, to investigate with `signifinder` the RNA sequencing samples of Ovarian Cancer (OVC) tumors available at The Cancer Genome Atlas (TCGA) project. RNA-Seq OVC data are retrieved from TCGA database using `curatedTCGAData` package. That returns a `MultiAssayExperiment`, from which we extract a `SummarizedExperiment`. Then, data are normalized using the `betweenLaneNormalization` function from the `EDASeq` package (version 2.32.0).

```{r}
library(curatedTCGAData)
OVmae <- curatedTCGAData(
    diseaseCode = "OV", assays = "RNASeqGene", version = "2.0.1", dry.run = FALSE)
OVse <- OVmae[[1]]
colnames(OVse) <- substr(colnames(OVse), 1, 12)
OVse <- OVse[,!duplicated(colnames(OVse))]

library(EDASeq)
assay(OVse) <- betweenLaneNormalization(as.matrix(assay(OVse)), which = "median")
names(assays(OVse)) <- "norm_expr"
```

# Compute Signatures
We look for signatures useful for OVC investigation. Hence, those signatures specifically developed in OVC or signatures that are available for all cancer types (pan-cancer).

```{r}
library(signifinder)
OVse <- multipleSign(dataset = OVse, nametype = "SYMBOL", inputType = "rnaseq",
                     tissue = c("ovary", "pan-tissue"))
```

# Visualize Scores
Figure 2A. We compute the correlation matrix of the signature scores and perform a hierarchical clustering to group similar correlations. For this step we use the `dist` function and the `hclust` function from the `stats` package (version 4.2.2). We then compute the silhouette index to find the best number of clusters and discuss the ones containing the higher correlation values (`NbClust` package, version 3.0.1).

```{r}
sign_cor <- correlationSignPlot(OVse)

OVsecD <- colData(OVse)
OVsecD01 <- t(vapply(OVsecD, signifinder:::.range01, double(nrow(OVsecD))))
OVsecD01cor <- cor(apply(OVsecD01, 1, as.numeric), 
                   use = "pairwise.complete.obs", method = "pearson")
OVsecD01cor <- OVsecD01cor[apply(OVsecD01cor, 1, function(x) !all(is.na(x))), 
                           apply(OVsecD01cor, 2, function(x) !all(is.na(x)))]
score_hclust <- hclust(dist(OVsecD01cor), method = "complete")

library(NbClust)
hclust_index <- NbClust(data = OVsecD01cor, min.nc = 4, max.nc = 11,
    distance = "euclidean", method = "complete", index = "silhouette")

group1 <- c("EMT_Mak", "Matrisome_Yuzhalin", "LipidMetabolism_Zheng", 
            "ECM_Chakravarthy_up", "ECM_Chakravarthy_down")
group2 <- c("IPS_Charoentong_MHC", "ExpandedImmune_Ayers", "ImmunoScore_Roh", "Tinflam_Ayers", 
            "ImmuneCyt_Davoli", "ImmunoScore_Hao", "IFN_Ayers", "ImmuneCyt_Rooney")
group3 <- c("Chemokines_Messina", "IPSOV_Shen", "IPS_Charoentong_EC", "IPS_Charoentong_CP")
group4 <- c("CellCycle_Davoli", "CellCycle_Lundberg", "CIN_Carter_70", "CIN_Carter_25", 
            "MitoticIndex_Yang", "DNArep_Kang")
```

Figure 2B. Generate the heatmap to visualize scores for the signatures inside group 1, 2, 3, and 4.

```{r}
cons_sign <- c("ConsensusOV_Chen_IMR", "ConsensusOV_Chen_DIF", "ConsensusOV_Chen_PRO", "ConsensusOV_Chen_MES")
OV_subtype <- sapply(1:ncol(OVse), function(x){
  names(which.max(as.vector(as.data.frame(OVsecD)[x, cons_sign])))})
OV_subtype <- substring(OV_subtype, 18)
names(OV_subtype) <- colnames(OVse)

groups <- c(rep("group1", length(group1)), rep("group2", length(group2)),
            rep("group3", length(group3)), rep("group4", length(group4)))

heatmapSignPlot(data = OVse, whichSign = c(group1, group2, group3, group4), 
                clusterBySign = cons_sign, sampleAnnot = OV_subtype, 
                splitBySampleAnnot = T, row_split = groups)
```

Figure 2C. Generate the ridge plots to visualize scores for the signatures inside group 1, 2, 3, and 4.

```{r}
ridgelineSignPlot(data = OVse, groupByAnnot = OV_subtype, whichSign = group1)
ridgelineSignPlot(data = OVse, groupByAnnot = OV_subtype, whichSign = group2)
ridgelineSignPlot(data = OVse, groupByAnnot = OV_subtype, whichSign = group3)
ridgelineSignPlot(data = OVse, groupByAnnot = OV_subtype, whichSign = group4)
```

Figure 2D. Plot survival curves for EMT_Miow_Mesenchymal signature.

```{r}
OVmaecD <- colData(OVmae)
os <- NULL
os_indx <- !is.na(OVmaecD$days_to_death)
os[os_indx] <- OVmaecD$days_to_death[os_indx]
os_indx <- !is.na(OVmaecD$days_to_last_followup)
os[os_indx] <- OVmaecD$days_to_last_followup[os_indx]
names(os) <- OVmaecD@rownames

status <- OVmaecD$vital_status
names(status) <- OVmaecD@rownames

mysurvData <- cbind(os, status)
survivalSignPlot(data = OVse, survData = mysurvData, 
                 whichSign = "EMT_Miow_Mesenchymal", cutpoint = "optimal")
```

Figure 2E. Generate a scatterplot of scores and a barplot of their density distribution for ASC_Smith signature.

```{r}
oneSignPlot(data = OVse, whichSign = "ASC_Smith", statistics = "quantiles")
```

Figure 2F. Generate a boxplot of the ASC_Smith signature scores divided by OV subtype.

```{r}
box_data <- data.frame(OV_subtype = OV_subtype, score = colData(OVse)[,"ASC_Smith"])
my_comparisons <- list( c("PRO", "DIF"), c("PRO", "IMR"), c("PRO", "MES") )

library(ggpubr)
ggboxplot(box_data, "OV_subtype", "score", fill = "OV_subtype") +
    labs(title = "ASC_Smith", x = "", y = "score") + 
    stat_compare_means(comparisons = my_comparisons, method = "t.test")
```

Additional Figure 1. Generate the evaluation plot for all the signatures computed.

```{r}
evaluationSignPlot(data = OVse)
```