---
title: "SC_workflow"
author: "Stefania Pirrotta"
output: html_document
---

# Prepare Data
In this workflow, we present all the analyses, along with the code, to investigate with `signifinder` the single-cell RNA sequencing samples of the glioblastoma dataset of Darmanis et al. (GEO ID: GSE84465, GEO ID: GSE84465, Darmanis, S. et al. Single-Cell RNA-Seq Analysis of Infiltrating Neoplastic Cells at the Migrating Front of Human Glioblastoma. Cell Rep 21, 1399â€“1410 (2017)). We download the dataset using the `TMExplorer` R package (version 1.6.2). The `SingleCellExperiment` object provides the raw count expression values.

```{r}
library(TMExplorer)
GBsce <- queryTME(geo_accession = "GSE84465")
GBsce <- GBsce[[1]]
colnames(colData(GBsce)) <- "cell_type"
GBsce@assays@data@listData$counts <- as.matrix(GBsce@assays@data@listData$counts)
GBsce <- GBsce[which(apply(assay(GBsce) > 0, 1, any)),]

library(GEOquery)
GBsce_info <- getGEO(GEO = "GSE84465", GSEMatrix = TRUE)
GBsce_info <- GBsce_info[[1]]@phenoData@data
patient <- substring(GBsce_info$characteristics_ch1.4, 13)
location <- substring(GBsce_info$characteristics_ch1.3, 9)
location[location=="Tumor"] <- "Core"
GBsce$patient <- patient
GBsce$location <- location
```

We filter genes and cells following the `scone` workflow (version 1.22.0). Then, we select two samples, one containing the core cells and the other the peripheral cells of the tumor mass from the BT_S2 patient, and we remove the zero-expression genes.

```{r}
library(scone)
data(housekeeping)
hk <- intersect(housekeeping$V1, rownames(assay(GBsce)))

num_reads <- quantile(assay(GBsce)[assay(GBsce) > 0])[4]
num_cells <- 0.25 * ncol(GBsce)
keep <- which(DelayedArray::rowSums(assay(GBsce) >= num_reads) >= num_cells)
mfilt <- metric_sample_filter(assay(GBsce), gene_filter = keep,
                              pos_controls = rownames(GBsce) %in% hk,
                              zcut = 3, mixture = FALSE, plot = FALSE)
mfilt <- !apply(simplify2array(mfilt[!is.na(mfilt)]), 1, any)

num_reads <- quantile(assay(GBsce)[assay(GBsce) > 0])[4]
num_cells <- 5
is_quality <- which(DelayedArray::rowSums(assay(GBsce) >= num_reads) >= num_cells)

GBsce <- GBsce[is_quality, mfilt]

GBsce <- GBsce[, GBsce$patient=="BT_S2"]
GBsce <- GBsce[which(apply(assay(GBsce) > 0, 1, any)),]
```

Then, using `scone` workflow, we apply the best suggested normalization method.

```{r}
expr <- assay(GBsce)
negcon = intersect(rownames(expr), hk)
my_scone <- SconeExperiment(expr, negcon_ruv = rownames(expr) %in% negcon)

EFF_FN = function (ei) {
  sums = colSums(ei > 0)
  eo = t(t(ei)*sums/mean(sums))
  return(eo) }

scaling <- list(none = identity, eff = EFF_FN, sum = SUM_FN, tmm = TMM_FN,
                uq = UQ_FN, fq = FQT_FN, psi = PSINORM_FN, deseq = DESEQ_FN)
BiocParallel::register(BiocParallel::SerialParam())
my_scone <- scone(my_scone, scaling = scaling, run = TRUE, eval_kclust = 2:6,
                  stratified_pam = FALSE, return_norm = "in_memory",
                  zero = "postadjust", k_qc = 0)

out_norm = get_normalized(my_scone, method = rownames(get_params(my_scone))[1])
```

Additionally, to reduce the large number of false zeros, we run ALRA on the normalized dataset (PMID: 35017482).

```{r}
source("https://raw.githubusercontent.com/KlugerLab/ALRA/master/alra.R")
sample_names <- colnames(out_norm)
k_choice <- choose_k(t(out_norm))
result.completed <- alra(t(out_norm), k = k_choice$k)
out_norm_completed <- result.completed[[3]]
rownames(out_norm_completed) <- sample_names

assays(GBsce, withDimnames = FALSE)[["norm_expr"]] <- t(out_norm_completed)
```

# Compute Signatures
We compute all the signatures for brain and pan-tissue cancers that are available in signifinder.

```{r}
library(signifinder)
GBsce <- multipleSign(dataset = GBsce, nametype = "SYMBOL", inputType = "rnaseq",
                      tissue = c("brain", "pan-tissue"))
```

We performed t-distributed stochastic neighbor embedding (t-SNE) using the top 50 components of the PCA (`scater` package, version 1.26.1). Further, to have sizable cell groups for signature score comparisons, we keep only the cell types that have greater than 20 cells present in both core and peripheral samples. Hence, immune cells, neoplastic, and oligodendrocyte precursor cells are considered for further analysis.

```{r}
library(scater)
GBsce <- runPCA(GBsce, ncomponents = 50, ntop = 1000, scale = TRUE,
                BSPARAM = BiocSingular::RandomParam(), exprs_values = "norm_expr")
GBsce <- runTSNE(GBsce, dimred = "PCA", external_neighbors = TRUE,
                 BNPARAM = BiocNeighbors::AnnoyParam())

GBsce <- GBsce[, -which(
    GBsce$cell_type %in% c("Oligodendrocyte", "Neuron", "Astocyte", "Vascular"))]

GBsce$cell_loc <- paste(GBsce$cell_type, "-", GBsce$location)
```

# Visualize Scores
Generate the color palettes.

```{r}
mypal <- RColorBrewer::brewer.pal(n = 11, name = "Spectral")[c(1:4, 6, 8:11)]
myPalette <- colorRampPalette(rev(mypal))
discrete_palette <- c(
    "Immune cell - Periphery" = "#F5C566", "Immune cell - Core" = "#FB8B24",
    "Neoplastic - Periphery" = "#D08A98", "Neoplastic - Core" = "#CF0026",
    "OPC - Periphery" = "#85D5BC", "OPC - Core" = "#034732")
```

Figure 3A. Plot the t-SNE colored by cell type and cell location.

```{r}
plotTSNE(GBsce, colour_by = "cell_loc", point_size = 1.7, point_alpha = 1) + 
    scale_color_manual(values = discrete_palette) +
    labs(color = "Tissue - Localization") + 
    theme(legend.key.size = unit(.3, units = "cm"), 
          text = element_text(size = 7), axis.text=element_text(size = 7), 
          legend.text = element_text(size = 8), legend.title = element_text(size = 8))
```

Figures 3B, C, D, E. Plot t-SNE colored by scores of Hypoxia_Buffa, CellCycle_Davoli, ECM_Chakravarthy_up and ECM_Chakravarthy_down and the corresponding boxplots.

```{r}
lapply(c("Hypoxia_Buffa", "CellCycle_Davoli"), function(i){
    plotTSNE(GBsce, colour_by = i, point_size = 1) + 
        labs(colour = "score") + 
        scale_colour_gradientn(colours = myPalette(100)) + 
        ggtitle(i) + 
        theme(legend.key.size = unit(0.2, units = "cm"), 
              text = element_text(size = 7), axis.text=element_text(size = 7))})

lapply(c("ECM_Chakravarthy_up", "ECM_Chakravarthy_down"), function(i){
    a <- c(colData(GBsce)[,"ECM_Chakravarthy_up"],
           colData(GBsce)[,"ECM_Chakravarthy_down"])
    plotTSNE(GBsce, colour_by = i, point_size=1) + 
        labs(colour = "score") + 
        scale_colour_gradientn(colours = myPalette(100), limits = c(min(a), max(a))) + 
        ggtitle(i) + 
        theme(legend.key.size = unit(0.2, units = "cm"), 
              text = element_text(size = 7), axis.text=element_text(size = 7))})

lapply(c("Hypoxia_Buffa", "CellCycle_Davoli", "ECM_Chakravarthy_up", 
         "ECM_Chakravarthy_down"), function(i){
    tmp <- GBsce
    ggplot(mapping = aes(
        x = tmp$cell_type, y = colData(tmp)[,i], dodge = tmp$location, 
        fill = interaction(tmp$cell_type, tmp$location))) + 
    geom_boxplot(alpha = 1, outlier.size = .8) +
    scale_fill_manual(values = c(
        "Immune cell.Periphery" = "#F5C566", "Immune cell.Core" = "#FB8B24",
        "Neoplastic.Periphery" = "#D08A98", "Neoplastic.Core" = "#CF0026",
        "OPC.Periphery" = "#85D5BC", "OPC.Core" = "#034732"), 
        breaks = c("Immune cell.Core", "Immune cell.Periphery",
                   "Neoplastic.Core", "Neoplastic.Periphery",
                   "OPC.Core", "OPC.Periphery")) +
    labs(x = "", title = i, y = "score", fill = "Tissue - Localization") +
    theme_minimal() +
    theme(text = element_text(size = 7), axis.text = element_text(size = 7), 
          legend.position = "none")})
```

Figures 3F, G, H, I. Plot t-SNE colored by scores of Hypoxia_Buffa, CellCycle_Davoli and VEGF_Hu for the neoplastic cells and the corresponding boxplots.

```{r}
sign_neo <- c("Hypoxia_Buffa", "VEGF_Hu", "CellCycle_Davoli")

ridgelineSignPlot(data = GBsce, groupByAnnot = GBsce$cell_loc, 
                  selectByAnnot = c("Neoplastic - Periphery", "Neoplastic - Core"), 
                  whichSign = sign_neo, alpha = 0.8) + 
    scale_fill_manual(values = c("Neoplastic - Periphery" = "#D08A98", 
                                 "Neoplastic - Core" = "#CF0026")) + 
    theme_classic(6) + labs(y="") + 
    theme(legend.position = "none", 
          text = element_text(size = 7), axis.text=element_text(size = 7))

lapply(sign_neo, function(i){
    tmp <- GBsce
    colData(tmp)[tmp$cell_type!="Neoplastic",i] <- NA
    plotTSNE(tmp, colour_by = i, point_size=0.4) + 
        labs(colour = "score") + 
        scale_colour_gradientn(colours = myPalette(100), na.value = "grey90") + 
        ggtitle(i) + 
        theme(legend.key.size = unit(0.2, units = "cm"), 
              text = element_text(size = 7), axis.text=element_text(size = 7))})

library(ggpubr)
lapply(sign_neo, function(i){
    tmp <- GBsce
    tmp <- tmp[, tmp$cell_type=="Neoplastic"]
    my_comparisons <- list( c("Periphery", "Core") )
    ggplot(mapping = aes(x = tmp$location, y = colData(tmp)[,i], 
                         fill = tmp$location, dodge = tmp$location)) + 
    geom_boxplot(alpha = 1, outlier.size = .8) +
    scale_fill_manual(values = c(
        "Periphery" = "#D08A98", "Core" = "#CF0026")) +
    labs(x="", title=i, y="score", fill="Tissue - Localization") +
    theme_minimal() +
    theme(text = element_text(size = 7), axis.text=element_text(size = 7), 
          legend.position = "none") + 
    stat_compare_means(comparisons = my_comparisons, method = "t.test")})
```

Figures 3J, K, L, M, N, O, P. Plot t-SNE colored by scores of IPS_Charoentong_CP, IPS_Charoentong_EC, IPS_Charoentong_MHC, IPS_Charoentong_SC, ImmunoScore_Roh and Chemokines_Messina for the immune cells and the corresponding boxplots.

```{r}
sign_imm <- c(
    "IPS_Charoentong_CP", "IPS_Charoentong_EC", "IPS_Charoentong_MHC", 
    "IPS_Charoentong_SC", "ImmunoScore_Roh", "Chemokines_Messina")

ridgelineSignPlot(data = GBsce, groupByAnnot = GBsce$cell_loc, 
                  selectByAnnot = c("Immune cell - Periphery", "Immune cell - Core"), 
                  whichSign = sign_imm, alpha = 0.8) + 
    scale_fill_manual(values = c("Immune cell - Periphery" = "#F5C566", 
                                 "Immune cell - Core" = "#FB8B24")) + 
    theme_classic(6) + labs(y="") + 
    theme(legend.position = "none", 
          text = element_text(size = 7), axis.text=element_text(size = 7))

lapply(sign_imm, function(i){
    tmp <- GBsce
    colData(tmp)[tmp$cell_type!="Immune cell",i] <- NA
    plotTSNE(tmp, colour_by = i, point_size=0.4) + 
        labs(colour = "score") + 
        scale_colour_gradientn(colours = myPalette(100), na.value = "grey90") + 
        ggtitle(i) + 
        theme(legend.key.size = unit(0.2, units = "cm"), 
              text = element_text(size = 7), axis.text=element_text(size = 7))})

lapply(sign_imm, function(i){
    tmp <- GBsce
    tmp <- tmp[, tmp$cell_type=="Immune cell"]
    my_comparisons <- list( c("Periphery", "Core") )
    ggplot(mapping = aes(x = tmp$location, y = colData(tmp)[,i], 
                         fill = tmp$location, dodge = tmp$location)) + 
    geom_boxplot(alpha = 1, outlier.size = .8) +
    scale_fill_manual(values = c(
        "Periphery" = "#F5C566", "Core" = "#FB8B24")) +
    labs(x="", title=i, y="score", fill="Tissue - Localization") +
    theme_minimal() +
    theme(text = element_text(size = 7), axis.text=element_text(size = 7), 
          legend.position = "none") + 
    stat_compare_means(comparisons = my_comparisons, method = "t.test")})
```

Additional Figure 2. Generate the evaluation plot for all the signatures computed.

```{r}
evaluationSignPlot(GBsce)
```

Additional Figure 3. Generate correlation matrices between signature scores of VEGF_Hu, Hypoxia_Buffa and CellCycle_Davoli computed on the neoplastic cells and between signature scores of Chemokines_Messina, ImmunoScore_Roh and IPS_Charoentong computed on the immune cells.

```{r}
pcor_n <- correlationSignPlot(data = GBsce, whichSign = sign_neo, 
    sampleAnnot = GBsce$cell_type, selectByAnnot = "Neoplastic")
pcor_i <- correlationSignPlot(data = GBsce, whichSign = sign_imm, 
    sampleAnnot = GBsce$cell_type, selectByAnnot = "Immune cell")
```

Additional Figure 4. Generate the scatterplot of CellCycle_Davoli scores and CIN_Carter_70 scores computed on the neoplastic cells.

```{r}
ggplot(mapping = aes(x = GBsce$CellCycle_Davoli[GBsce$cell_type=="Neoplastic"], 
                     y = GBsce$CIN_Carter_70[GBsce$cell_type=="Neoplastic"])) + 
    geom_point() + labs(x = "CellCycle_Davoli", y = "CIN_Carter_70") + 
    theme_bw() + theme(text = element_text(size = 15))
```
